src/Specular/Internal/Incremental/PriorityQueue.js

```diff
@@ -0,0 +1,88 @@
+// new :: forall a.
+//   EffectFn4
+//     (Optional Any)                 -- Optional.none
+//     (Field a Mutable Int)          -- Priority
+//     (Field a Mutable Boolean)      -- Is the entry present in this queue?
+//     (Field a Mutable (Optional a)) -- Next entry with the same priority
+//   (PQ a)
+exports.new = function(none, priorityField, presentField, nextField) {
+  return {
+    none: none,
+    priorityField: priorityField,
+    presentField: presentField,
+    nextField: nextField,
+
+    priorityHeads: [],
+    count: 0,
+
+    // TODO: we should track minPriority
+  };
+};
+
+// add :: forall a. EffectFn2 (PQ a) a Boolean
+exports.add = function(pq, node) {
+  if(node[pq.presentField]) {
+    return false;
+  }
+
+  node[pq.presentField] = true;
+  pq.count++;
+
+  var priority = node[pq.priorityField];
+  while(priority >= pq.priorityHeads.length) {
+    pq.priorityHeads.push(pq.none);
+  }
+
+  node[pq.nextField] = pq.priorityHeads[priority];
+  pq.priorityHeads[priority] = node;
+
+  return true;
+};
+
+// remove :: forall a. EffectFn2 (PQ a) a Boolean
+exports.remove = function(pq, node) {
+  if(!node[pq.presentField]) {
+    return false;
+  }
+
+  node[pq.presentField] = false;
+
+  var priority = node[pq.priorityField];
+
+  if (pq.priorityHeads[priority] === node) {
+    pq.priorityHeads[priority] = node[pq.nextField];
+  }
```

Comment:
What if the node is not a priority head? I think we should find its predecessor and do `pred[pq.nextField] = node[pq.nextFieled]`.

Reply:
Wow, good catch! This is super wrong. But turns out this funcion isn't even used. Will remove.